Objects/listobject.c

```diff
@@ -970,7 +1093,7 @@ list_extend_fast(PyListObject *self, PyObject *iterable)
     // an overflow on any relevant platform.
     assert(m < PY_SSIZE_T_MAX - n);
     if (self->ob_item == NULL) {
-        if (list_preallocate_exact(self, n) < 0) {
+        if (list_ensure_capacity(self, m, n) < 0) {
```

Comment:
IIRC, the idea behind `list_ensure_capacity` was to subsume these extra checks. i.e. to just write:

```
if (list_ensure_capacity(self, m, n) < 0) {
  return NULL;
}
...
PyList_SET_SIZE(self, m + n);
```

No `self->ob_item == NULL` check or  extra `list_resize` call.

See https://github.com/colesbury/nogil-3.12/blob/cedde4f5ec3759ad723c89d44738776f362df564/Objects/listobject.c#L1206-L1220